agent/internal/detect.go

```diff
@@ -71,11 +73,69 @@ func detectCPUs() ([]device.Device, error) {
 	}
 }
 
+var detectMIGEnabled = []string{
+							"nvidia-smi", "--query-gpu=mig.mode.current", "--format=csv,noheader"}
+var detectNvidiaDevices = []string{"nvidia-smi", "-L"} // Lists both GPUs and MIG instances
+var detectMIGRegExp = regexp.MustCompile(`(?P<dev>MIG \S+).+\(UUID.+(?P<uuid>MIG.+)\)`)
+
 var detectGPUsArgs = []string{"nvidia-smi", "--query-gpu=index,name,uuid", "--format=csv,noheader"}
 var detectGPUsIDFlagTpl = "--id=%v"
 
+// detect if MIG is enabled and if there are instances configured.
+func detectMigInstances(visibleGPUs string) ([]device.Device, error) {
+	// Fail fast if MIG isn't even enabled
+	// #nosec G204
+	cmd := exec.Command(detectMIGEnabled[0], detectMIGEnabled[1:]...)
+	out, err := cmd.Output()
+	if execError, ok := err.(*exec.Error); ok && execError.Err == exec.ErrNotFound {
+		return nil, nil
+	} else if err != nil {
+		log.WithError(err).WithField("output", string(out)).Warnf(
+			"error while executing nvidia-smi to detect MIG mode")
+		return nil, nil
+	}
+	if !strings.HasPrefix(string(out), "Enabled") {
+		return nil, nil
+	}
+
+	// #nosec G204
+	cmd = exec.Command(detectNvidiaDevices[0], detectNvidiaDevices[1:]...)
+	out, err = cmd.Output()
+	if err != nil {
+		log.WithError(err).WithField("output", string(out)).Warnf(
+			"error while executing nvidia-smi to detect MIG instances")
+		return nil, nil
+	}
+
+	devices := make([]device.Device, 0)
+	deviceIndex := 0
+
+	scanner := bufio.NewScanner(strings.NewReader(string(out)))
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if detectMIGRegExp.MatchString(line) {
+			matches := detectMIGRegExp.FindStringSubmatch(line)
+			if len(matches) != 3 {
+				continue
+			}
+			brand := matches[1]
+			uuid := matches[2]
+			devices = append(devices,
+				device.Device{ID: deviceIndex, Brand: brand, UUID: uuid, Type: device.GPU})
+			deviceIndex++
+		}
+	}
+	return devices, err
```

Comment:
i prefer explicit `return devices, nil`. save the next reader of needing to reparse to make sure it's always nil.